File: zig/anagram/anagram.zig
const std = @import("std");
const mem = std.mem;
const ascii = std.ascii;

fn countLetters(word: []const u8) [26]u8 {
    var counts = [_]u8{0} ** 26;
    for (word) |c| {
        if (ascii.isAlphabetic(c)) counts[ascii.toLower(c) - 'a'] += 1;
    }

    return counts;
}
fn isAnagram(w1: []const u8, w2: []const u8) bool {
    if (ascii.eqlIgnoreCase(w1, w2)) return false;

    const c1 = countLetters(w1);
    const c2 = countLetters(w2);

    return std.mem.eql(u8, &c1, &c2);
}

/// Returns the set of strings in `candidates` that are anagrams of `word`.
/// Caller owns the returned memory.
pub fn detectAnagrams(
    allocator: mem.Allocator,
    word: []const u8,
    candidates: []const []const u8,
) !std.BufSet {
    var ans = std.BufSet.init(allocator);
    for (candidates) |candidate| {
        if (isAnagram(word, candidate)) try ans.insert(candidate);
    }
    return ans;
}

File: cpp/anagram/anagram.cpp
#include "anagram.h"

#include <cctype>

namespace anagram {
map<char, int> anagram::count(string s) {
    map<char, int> ans;
    for (char ch : s) {
        if (isalpha(ch)) ans[tolower(ch)]++;
    }
    return ans;
}

bool anagram::is_equal(string s) {
    if (s.size() != this->_word.size()) return false;
    for (size_t i = 0; i < s.size(); i++) {
        if (tolower(s[i]) != tolower(this->_word[i])) return false;
    }
    return true;
}

bool anagram::is_anagram(string s) {
    return count(s) == this->_count && !is_equal(s);
}

anagram::anagram(string s) {
    this->_word = s;
    this->_count = count(s);
}

vector<string> anagram::matches(vector<string> words) {
    vector<string> ans;
    for (string word : words) {
        if (is_anagram(word)) ans.emplace_back(word);
    }
    return ans;
}
}  // namespace anagram
